File: 2024/Day5/main.py
from __future__ import annotations
from functools import cmp_to_key
SAMPLE_DATA: bool = False
if SAMPLE_DATA:
    filename = "sample_data.txt"
else:
    filename = "data.txt"


with open(filename, 'r') as f:
    data = [l.strip() for l in f]

class OrderingRule:
    pageX: int
    pageY: int

    def __init__(self, x: int, y: int):
        self.pageX = x
        self.pageY = y

    @staticmethod
    def from_string(s: str):
        x, y = s.split("|")
        return OrderingRule(int(x), int(y))
    
    def is_update_valid(self, update: Update):
        try:
            return update.pages.index(self.pageX) < update.pages.index(self.pageY)
        except ValueError:
            return True
    
    def __repr__(self):
        return f"OrderingRule: {self.pageX} | {self.pageY}"

class Update:
    pages: list[int]

    def __init__(self, pages: list[int]):
        self.pages = pages

    @staticmethod
    def from_string(s: str):
        return Update([int(x) for x in s.split(',')])
    
    def center_page(self):
        return self.pages[len(self.pages) // 2]
    
    def fix_update(self, rules: list[OrderingRule]):
        def comparer(lhs: int, rhs: int):
            for r in rules:
                if r.pageX == lhs and r.pageY == rhs:
                    return -1
                if r.pageX == rhs and r.pageY == lhs:
                    return 1
            raise ValueError(f"Invalid ordering: {lhs} {rhs}")
        self.pages.sort(key=cmp_to_key(comparer))
    
    def __repr__(self):
        return f"Update: {self.pages}"
    

divider_idx = data.index('')
rules = [OrderingRule.from_string(x) for x in data[:divider_idx]]
updates = [Update.from_string(x) for x in data[divider_idx+1:]]
part1 = 0
part2 = 0
for up in updates:
    if all((rule.is_update_valid(up) for rule in rules)):
        part1 += up.center_page()
    else:
        up.fix_update(rules)
        part2 += up.center_page() 
print(f'{part1} {part2}')

File: 2024/Day5/main.cpp
#include <algorithm>
#include <charconv>
#include <fstream>
#include <ranges>
#include <string>
#include <string_view>
#include <vector>
#include <cstdio>
#include <unordered_map>
#include <numeric>
#include <print>
#include <cmath>

namespace r = std::ranges;
namespace v = std::ranges::views;
using namespace std::string_view_literals;

constexpr bool SAMPLE_DATA = false;

std::vector<std::string> read_file(std::string_view filename) {
    std::ifstream file(filename.data());
    std::vector<std::string> lines;
    std::string line;
    while (std::getline(file, line)) {
        lines.push_back(line);
    }
    return lines;
}

struct ordering_rule;

struct update {
    std::vector<int64_t> pages;

    update(r::range auto&& range) {
        auto r = range | v::split(',');
        for (const auto& el : r) {
            int64_t page;
            std::from_chars(el.data(), el.data() + el.size(), page);
            pages.push_back(page);
        }
    }

    int64_t center_page() const {
        return pages[pages.size() / 2];
    }

    void fix_update(const std::vector<ordering_rule>& rule);
};

struct ordering_rule {
    int64_t page_x;
    int64_t page_y;

    ordering_rule(r::range auto&& range) {
        auto r = range | v::split('|');
        auto it = r.begin();
        std::from_chars((*it).data(), (*it).data() + (*it).size(), page_x);
        ++it;
        std::from_chars((*it).data(), (*it).data() + (*it).size(), page_y);
    }

    bool is_update_valid(const update& u) const {
        auto page_x_idx = r::find(u.pages, page_x);
        auto page_y_idx = r::find(u.pages, page_y);
        if (page_x_idx == u.pages.end() || page_y_idx == u.pages.end()) {
            return true;
        }
        return std::distance(page_x_idx, page_y_idx) > 0;
    }
};

void update::fix_update(const std::vector<ordering_rule>& rules) {
    r::sort(pages, [&rules](const auto& lhs, const auto& rhs) {
        for (const auto& r : rules)   {
            if (r.page_x == lhs && r.page_y == rhs) {
                return false;
            }
            if (r.page_x == rhs && r.page_y == lhs) {
                return true;
            }
        } 
        throw std::runtime_error("Invalid ordering rule");
    });
}

int main() {
    auto lines = read_file(SAMPLE_DATA ? "sample_data.txt"sv : "data.txt"sv);
    std::vector<ordering_rule> rules;
    std::vector<update> updates;
    bool parsing_rules = true;
    for (const auto& line : lines) {
        if (line.empty()) {
            parsing_rules = false;
            continue;
        }
        if (parsing_rules) {
            rules.emplace_back(line);
        } else {
            updates.emplace_back(line);
        }
    }

    int64_t part1 = 0;
    int64_t part2 = 0;

    for (auto& u : updates) {
        if (r::all_of(rules, [&u](const auto& r) {
            return r.is_update_valid(u);
        })) {
            part1 += u.center_page();
        } else {
            u.fix_update(rules);
            part2 += u.center_page();
        }
    }

    std::println("{} {}", part1, part2);
}